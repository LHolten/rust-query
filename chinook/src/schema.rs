use std::collections::HashMap;

use rust_query::{schema, Client, NoTable, Null, Prepare};

#[schema]
#[version(0..3)]
enum Schema {
    Album {
        #[version(1..)]
        something: String,
        title: String,
        artist: Artist,
    },
    Artist {
        name: String,
    },
    Customer {
        #[version(..2)]
        phone: Option<String>,
        #[version(2..)]
        phone: Option<i64>,
        first_name: String,
        last_name: String,
        company: Option<String>,
        address: String,
        city: String,
        state: Option<String>,
        country: String,
        postal_code: Option<String>,
        fax: Option<String>,
        email: String,
        support_rep: Employee,
    },
    #[version(1..)]
    #[unique(employee, artist)]
    ListensTo {
        employee: Employee,
        artist: Artist,
    },
    Employee {
        last_name: String,
        first_name: String,
        title: Option<String>,
        reports_to: Option<Employee>,
        birth_day: Option<String>,
        hire_date: Option<String>,
        address: Option<String>,
        city: Option<String>,
        state: Option<String>,
        country: Option<String>,
        postal_code: Option<String>,
        phone: Option<String>,
        fax: Option<String>,
        email: String,
    },
    #[version(..2)]
    Genre {
        name: String,
    },
    #[version(2..)]
    #[unique(name)]
    #[create_from(Genre)]
    Genre {
        name: String,
    },
    Invoice {
        customer: Customer,
        invoice_date: String,
        billing_address: Option<String>,
        billing_city: Option<String>,
        billing_state: Option<String>,
        billing_country: Option<String>,
        billing_postal_code: Option<String>,
        total: f64,
    },
    InvoiceLine {
        invoice: Invoice,
        track: Track,
        unit_price: f64,
        quantity: i64,
    },
    #[version(..2)]
    MediaType {
        name: String,
    },
    Playlist {
        name: String,
    },
    #[unique(playlist, track)]
    PlaylistTrack {
        playlist: Playlist,
        track: Track,
    },
    Track {
        name: String,
        album: Album,
        #[version(..2)]
        media_type: MediaType,
        #[version(2..)]
        media_type: String,
        genre: Genre,
        composer: Option<String>,
        #[version(2..)]
        composer_table: Option<Composer>,
        milliseconds: i64,
        bytes: i64,
        unit_price: f64,
        #[version(2..)]
        byte_price: f64,
    },
    #[version(2..)]
    Composer {
        name: String,
    },
}

pub fn migrate() -> (Client, v2::Schema) {
    let artist_title = HashMap::from([("a", "b")]);
    let m = Prepare::open("test.db");
    let m = m.create_batch(&[
        include_str!("../Chinook_Sqlite.sql"),
        include_str!("../migrate.sql"),
    ]);

    m.migrate(|_schema| v1::up::Schema {
        album: |row, album| {
            let artist = row.get(album.artist.name);
            Box::new(v1::up::AlbumMigration {
                something: artist_title.get(&*artist).copied().unwrap_or("unknown"),
                // new_title: album.title,
            })
        },
    })
    .migrate(|_schema| v2::up::Schema {
        customer: |row, customer| {
            Box::new(v2::up::CustomerMigration {
                phone: row.get(customer.phone).and_then(|x| x.parse::<i64>().ok()),
            })
        },
        track: |row, track| {
            Box::new(v2::up::TrackMigration {
                media_type: track.media_type.name,
                composer_table: Null::<NoTable>::default(),
                byte_price: row.get(track.unit_price) / row.get(track.bytes) as f64,
            })
        },
        genre: |_row, genre| Some(Box::new(v2::up::GenreMigration { name: genre.name })),
    })
    .finish()
}

#[cfg(test)]
mod tests {
    use super::*;
    use rust_query::expect;

    #[test]
    fn backwards_compat() {
        v0::assert_hash(expect!["38f654ce24217792"]);
        v1::assert_hash(expect!["d9962ef27f0ea2e8"]);
    }
}
