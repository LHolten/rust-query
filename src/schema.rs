use std::collections::HashMap;

use heck::{ToSnekCase, ToUpperCamelCase};
use proc_macro2::TokenStream;
use quote::{format_ident, quote, ToTokens};

use crate::{
    new_query, pragma,
    value::{Const, Value},
};

pub fn generate() -> String {
    let mut output = TokenStream::new();
    output.extend(quote! {
        use rust_query::{value::Db, Builder, HasId, Table};
        use std::marker::PhantomData;
    });

    let tables = new_query(|q| {
        let table = q.flat_table(pragma::TableList);
        q.filter(table.schema.eq(Const("main".to_owned())));
        q.filter(table.r#type.eq(Const("table".to_owned())));
        q.filter(table.name.eq(Const("sqlite_schema".to_owned())).not());
        q.into_vec(u32::MAX, |row| row.get(q.select(&table.name)))
    });

    for table in &tables {
        let columns = new_query(|q| {
            let table = q.flat_table(pragma::TableInfo(table.to_owned()));

            q.into_vec(u32::MAX, |row| {
                let name = row.get(q.select(table.name));
                let typ = row.get(q.select(table.r#type));
                let pk = row.get(q.select(table.pk)) != 0;
                let notnull = row.get(q.select(table.notnull)) != 0;
                (name, typ, pk, notnull)
            })
        });

        let fks: HashMap<_, _> = new_query(|q| {
            let fk = q.flat_table(pragma::ForeignKeyList(table.to_owned()));
            q.into_vec(u32::MAX, |row| {
                (row.get(q.select(fk.from)), row.get(q.select(fk.table)))
            })
        })
        .into_iter()
        .collect();

        let mut ids = columns.iter().filter(|x| x.2);
        let mut has_id = ids.next().cloned();
        if ids.next().is_some() {
            has_id = None;
        }

        let make_field = |name: &str| {
            let mut normalized = &*name.to_snek_case();
            if fks.contains_key(name) {
                normalized = normalized.trim_end_matches("_id");
            }
            format_ident!("{normalized}")
        };

        let make_type = |typ: &str, name: &str| {
            Some(match typ {
                "INTEGER" => {
                    if let Some(other) = fks.get(name) {
                        let other_ident = format_ident!("{}", other.to_upper_camel_case());
                        other_ident.to_token_stream()
                    } else {
                        quote!(i64)
                    }
                }
                x if x.starts_with("NVARCHAR") => quote!(String),
                _ => return None,
            })
        };

        let defs = columns.iter().filter_map(|(name, typ, pk, notnull)| {
            if has_id.is_some() && *pk {
                return None;
            }
            let mut typ = make_type(typ, name)?;
            if !notnull {
                typ = quote!(Option<#typ>);
            }
            let ident = make_field(name);
            Some(quote!(pub #ident: Db<'t, #typ>))
        });

        let inits = columns.iter().filter_map(|(name, typ, pk, _notnull)| {
            if has_id.is_some() && *pk {
                return None;
            }
            make_type(typ, name)?;
            let ident = make_field(name);
            Some(quote!(#ident: f.col(#name)))
        });

        let table_ident = format_ident!("{}", table.to_upper_camel_case());
        let dummy_ident = format_ident!("{}Dummy", table.to_upper_camel_case());

        let has_id = has_id.as_ref().map(|(name, _typ, _pk, _notnull)| {
            quote!(
                impl HasId for #table_ident {
                    const ID: &'static str = #name;
                    const NAME: &'static str = #table;
                }
            )
        });

        output.extend(quote! {
            pub struct #table_ident;

            pub struct #dummy_ident<'t> {
                _phantom: PhantomData<dyn Fn(&'t ()) -> &'t ()>,
                #(#defs,)*
            }

            impl Table for #table_ident {
                type Dummy<'t> = #dummy_ident<'t>;

                fn name(&self) -> String {
                    #table.to_owned()
                }

                fn build(f: Builder<'_>) -> Self::Dummy<'_> {
                    #dummy_ident {
                        _phantom: PhantomData,
                        #(#inits,)*
                    }
                }
            }

            #has_id
        })
    }

    output.to_string()
}
