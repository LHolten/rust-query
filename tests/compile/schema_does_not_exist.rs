use rust_query::migration::schema;

#[schema(Schema)]
#[version(1..=1)]
pub mod vN {
    pub struct Foo;
    #[version(1..)]
    #[from(Foo)]
    pub struct FooNext;
}

pub fn main() {}
