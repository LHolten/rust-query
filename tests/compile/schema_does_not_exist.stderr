error: the previous schema does not exists
 --> tests/compile/schema_does_not_exist.rs:8:12
  |
8 |     #[from(Foo)]
  |            ^^^
